WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationViewModel.swift

```diff
@@ -0,0 +1,78 @@
+import Foundation
+import Yosemite
+
+/// View model for `ProductSharingMessageGenerationView`
+final class ProductSharingMessageGenerationViewModel: ObservableObject {
+    let viewTitle: String
+
+    var generateButtonTitle: String {
+        messageContent.isEmpty ? Localization.generate : Localization.regenerate
+    }
+
+    var generateButtonImageName: String {
+        messageContent.isEmpty ? "sparkles" : "arrow.counterclockwise"
+    }
+
+    @Published var messageContent: String = ""
+    @Published private(set) var generationInProgress: Bool = false
+    @Published private(set) var errorMessage: String?
+
+    private let siteID: Int64
+    private let url: String
+    private let stores: StoresManager
+
+    init(siteID: Int64,
+         productName: String,
+         url: String,
+         stores: StoresManager = ServiceLocator.stores) {
+        self.siteID = siteID
+        self.url = url
+        self.stores = stores
+        self.viewTitle = String.localizedStringWithFormat(Localization.title, productName)
+    }
+
+    @MainActor
+    func generateShareMessage() async {
+        // TODO: Analytics
+        errorMessage = nil
+        generationInProgress = true
+        do {
+            self.messageContent = try await withCheckedThrowingContinuation { continuation in
+                stores.dispatch(ProductAction.generateProductSharingMessage(siteID: siteID,
+                                                                            url: url,
+                                                                            languageCode: Locale.current.identifier,
+                                                                            completion: { result in
+                    continuation.resume(with: result)
+                }))
```

Comment:
Super nit/Suggestion: Consider extracting this to a private async function. Just to keep the continuation dealing code out for readability.

Reply:
👍  Updated in 15fdd450a78f685d0e5f0f0dc60355659f238917.